File: app/src/main/java/com/moggot/findmycarlocation/MapViewModel.java
package com.moggot.findmycarlocation;

import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;

import com.google.android.gms.maps.model.LatLng;
import com.moggot.findmycarlocation.data.model.route.Path;
import com.moggot.findmycarlocation.domain.LocationInteractor;
import com.moggot.findmycarlocation.domain.MainInteractor;
import com.moggot.findmycarlocation.domain.MapInteractor;
import com.moggot.findmycarlocation.retry.RetryManager;

import javax.inject.Inject;

public class MapViewModel extends BaseViewModel {

    private final MutableLiveData<Path> routeData = new MutableLiveData<>();

    @NonNull
    private final MapInteractor mapInteractor;
    @NonNull
    private final LocationInteractor locationInteractor;
    @NonNull
    private final MainInteractor mainInteractor;
    @NonNull
    private final RetryManager retryManager;

    @Inject
    public MapViewModel(@NonNull MapInteractor mapInteractor,
                        @NonNull LocationInteractor locationInteractor,
                        @NonNull MainInteractor mainInteractor,
                        @NonNull RetryManager retryManager) {
        this.mapInteractor = mapInteractor;
        this.locationInteractor = locationInteractor;
        this.mainInteractor = mainInteractor;
        this.retryManager = retryManager;
    }

    public void buildRoute() {
        compositeDisposable.add(locationInteractor.getLocation()
                .flatMap(location -> {
                    LatLng originLocation = new LatLng(location.getLatitude(), location.getLongitude());
                    return mapInteractor.getRoute(originLocation)
                            .filter(path -> !path.getRoutes().isEmpty())
                            .doOnSuccess(routeData::postValue);
                }).doOnError(throwable -> errorStatus.setValue(new ErrorStatus(ErrorStatus.LOCATION_ERROR)))
                .toObservable()
                .retryWhen(retryHandler -> retryHandler.flatMap(retryManager::observeRetries))
                .subscribe());
    }

    public LatLng drawCircle() {
        return mainInteractor.loadParkingData().getLocation();
    }

    public void retryCall() {
        retryManager.retry();
    }

    public void foundCar() {
        mainInteractor.markCarIsFound();
    }

    public MutableLiveData<Path> getRouteData() {
        return routeData;
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/map/ui/MapViewModel.kt
package com.moggot.findmycarlocation.map.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.asLiveData
import androidx.lifecycle.viewModelScope
import com.google.android.gms.maps.model.LatLng
import com.moggot.findmycarlocation.base.BaseViewModel
import com.moggot.findmycarlocation.base.data.Result
import com.moggot.findmycarlocation.base.data.Status
import com.moggot.findmycarlocation.data.model.route.Path
import com.moggot.findmycarlocation.map.domain.MapInteractor
import com.moggot.findmycarlocation.parking.domain.ParkingInteractor
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MapViewModel @Inject constructor(
    private val mapInteractor: MapInteractor,
    private val parkingInteractor: ParkingInteractor
) : BaseViewModel() {

    private val routeData = MutableLiveData<Result<Path>>()
    val parkingData = parkingInteractor.getParkingPlace().flowOn(Dispatchers.IO).asLiveData()
    lateinit var currentLocation: LatLng

    fun buildRoute(latitude: Double, longitude: Double) = viewModelScope.launch(Dispatchers.Main) {
        currentLocation = LatLng(latitude, longitude)
        mapInteractor.getRoute(latitude, longitude)
            .flowOn(Dispatchers.IO)
            .catch { Result(Status.ERROR, it) }
            .collect { routeData.value = Result(Status.SUCCESS, it) }
    }

    fun retryCall() {
        buildRoute(currentLocation.latitude, currentLocation.longitude)
    }

    fun foundCar() = viewModelScope.launch(Dispatchers.IO) {
        val parkingData = parkingData.value
        if (parkingData != null) {
            parkingInteractor.removeParkingPlace(parkingData.id)
        }
    }

    fun getRouteData(): LiveData<Result<Path>> = routeData
}
